src/solvers/dgsem_p4est/dg_2d_parabolic.jl

```diff
@@ -0,0 +1,406 @@
+
+        gradient_x_node = Ja11 * gradients_reference_1 + Ja21 * gradients_reference_2
+        gradient_y_node = Ja12 * gradients_reference_1 + Ja22 * gradients_reference_2
+
+        set_node_vars!(gradients_x, gradient_x_node, equations_parabolic, dg, i, j, element)
+        set_node_vars!(gradients_y, gradient_y_node, equations_parabolic, dg, i, j, element)
+      end
+
+    end
+  end
+
+  # Prolong solution to interfaces
+  @trixi_timeit timer() "prolong2interfaces" prolong2interfaces!(
+    cache_parabolic, u_transformed, mesh, equations_parabolic, dg.surface_integral, dg)
+
+  # Calculate interface fluxes for the gradient. This reuses P4est `calc_interface_flux!` along with a 
+  # specialization for AbstractEquationsParabolic.
+  @trixi_timeit timer() "interface flux" calc_interface_flux!(cache_parabolic.elements.surface_flux_values, 
+                                                              mesh, False(), # False() = no nonconservative terms
+                                                              equations_parabolic, dg.surface_integral, dg, 
+                                                              cache_parabolic)
+
+  # Prolong solution to boundaries
+  @trixi_timeit timer() "prolong2boundaries" prolong2boundaries!(
+    cache_parabolic, u_transformed, mesh, equations_parabolic, dg.surface_integral, dg)
+
+  # Calculate boundary fluxes
+  @trixi_timeit timer() "boundary flux" calc_boundary_flux_gradients!(
+    cache_parabolic, t, boundary_conditions_parabolic, mesh, equations_parabolic,
+    dg.surface_integral, dg)
+
+  # TODO: parabolic; mortars
+  @assert nmortars(dg, cache) == 0
+
+  # Calculate surface integrals
+  @trixi_timeit timer() "surface integral" begin
+    (; boundary_interpolation) = dg.basis
+    (; surface_flux_values) = cache_parabolic.elements
+    (; contravariant_vectors) = cache.elements
+
+    # Access the factors only once before beginning the loop to increase performance.
+    # We also use explicit assignments instead of `+=` to let `@muladd` turn these
+    # into FMAs (see comment at the top of the file).
+    factor_1 = boundary_interpolation[1,          1]
+    factor_2 = boundary_interpolation[nnodes(dg), 2]
+    @threaded for element in eachelement(dg, cache)
+      for l in eachnode(dg)
+        for v in eachvariable(equations_parabolic)
+
+          # Compute x-component of gradients
+
+          # surface at -x
+          normal_direction_x, _ = get_normal_direction(1, contravariant_vectors, 1, l, element)
+          gradients_x[v, 1,          l, element] += surface_flux_values[v, l, 1, element] * factor_1 * normal_direction_x
+
+          # surface at +x
+          normal_direction_x, _ = get_normal_direction(2, contravariant_vectors, nnodes(dg), l, element)
+          gradients_x[v, nnodes(dg), l, element] = (
+            gradients_x[v, nnodes(dg), l, element] + surface_flux_values[v, l, 2, element] * factor_2 * normal_direction_x)
```

Comment:
```suggestion
          gradients_x[v, nnodes(dg), l, element] += surface_flux_values[v, l, 2, element] * factor_2 * normal_direction_x
```
Same for the other lines below.

Reply:
If I remember correctly, we didn't do this because `@muladd` doesn't process `+=` correctly. 

https://github.com/SciML/MuladdMacro.jl#limitations

Reply:
Note that there's also a comment about this above the loop.

Reply:
Ah, thanks for the catch.